File: content/develop/use-cases/pub-sub/java-jedis/RedisPubSubHub.java
import redis.clients.jedis.Jedis;
import redis.clients.jedis.JedisPool;
import redis.clients.jedis.JedisPubSub;

import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Deque;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicLong;

/**
 * Redis-backed pub/sub hub helper.
 *
 * <p>Wraps PUBLISH / SUBSCRIBE / PSUBSCRIBE plus the introspection commands
 * (PUBSUB CHANNELS, PUBSUB NUMSUB, PUBSUB NUMPAT) into a small, named API
 * that:</p>
 *
 * <ul>
 *   <li>publishes JSON-encoded messages to a channel and counts how many
 *       subscribers Redis reported delivering to</li>
 *   <li>creates named in-process subscribers that own a dedicated
 *       {@link JedisPubSub} listener and a background dispatch thread,
 *       with each message wrapped as a {@link ReceivedMessage}</li>
 *   <li>tracks per-channel publish counters and per-subscriber received
 *       counts for the demo UI</li>
 * </ul>
 *
 * <p>Pub/sub has at-most-once delivery: a message that arrives while a
 * subscriber is disconnected is gone. If you need persistence or replay,
 * use Redis Streams instead.</p>
 */
public class RedisPubSubHub {

    /** Message shape delivered to every Subscription's buffer. */
    public static final class ReceivedMessage {
        final String channel;
        final String pattern; // null for exact-match subscriptions
        final Object payload;
        final long receivedAtMs;

        ReceivedMessage(String channel, String pattern, Object payload, long receivedAtMs) {
            this.channel = channel;
            this.pattern = pattern;
            this.payload = payload;
            this.receivedAtMs = receivedAtMs;
        }

        public String channel() { return channel; }
        public String pattern() { return pattern; }
        public Object payload() { return payload; }
        public long receivedAtMs() { return receivedAtMs; }

        public Map<String, Object> toMap() {
            Map<String, Object> out = new LinkedHashMap<>();
            out.put("channel", channel);
            out.put("pattern", pattern); // emits JSON null when no pattern
            out.put("payload", payload);
            out.put("received_at_ms", receivedAtMs);
            return out;
        }
    }

    /**
     * A named in-process subscriber bound to one or more channels or patterns.
     *
     * <p>Each {@code Subscription} owns its own {@link JedisPubSub} listener
     * running on a dedicated thread with a dedicated {@link Jedis}
     * connection. Subscriptions are independent: closing one does not
     * affect another, even if they share channels.</p>
     */
    public final class Subscription {
        private final String name;
        private final List<String> targets;
        private final boolean isPattern;
        private final int bufferSize;

        private final Deque<ReceivedMessage> buffer = new ArrayDeque<>();
        private final Object bufferLock = new Object();
        private long received = 0;
        private volatile boolean closed = false;

        private final Jedis jedis;
        private final JedisPubSub listener;
        private final Thread thread;

        Subscription(String name, List<String> targets, boolean isPattern, int bufferSize) {
            if (targets == null || targets.isEmpty()) {
                throw new IllegalArgumentException("Subscription requires at least one channel or pattern");
            }
            this.name = name;
            this.targets = List.copyOf(targets);
            this.isPattern = isPattern;
            this.bufferSize = bufferSize;

            // Each subscriber gets its own Jedis connection. Sharing one
            // connection across subscribers would couple their lifetimes —
            // unsubscribing one would close the channel for the others.
            this.jedis = pool.getResource();

            this.listener = new JedisPubSub() {
                @Override
                public void onMessage(String channel, String message) {
                    dispatch(channel, null, message);
                }

                @Override
                public void onPMessage(String pattern, String channel, String message) {
                    dispatch(channel, pattern, message);
                }
            };

            final String[] targetArr = this.targets.toArray(new String[0]);
            this.thread = new Thread(() -> {
                try {
                    if (isPattern) {
                        jedis.psubscribe(listener, targetArr);
                    } else {
                        jedis.subscribe(listener, targetArr);
                    }
                } catch (Exception ignored) {
                    // Connection closed by close(); thread exits.
                } finally {
                    try {
                        jedis.close();
                    } catch (Exception ignored) {
                    }
                }
            }, "pubsub-" + name);
            this.thread.setDaemon(true);
            this.thread.start();

            // jedis.subscribe / psubscribe send the SUBSCRIBE command
            // on its own connection. The command flushes immediately and
            // Redis registers the client as a subscriber before any
            // publish race can occur; tests that publish right after
            // returning from this constructor should still wait a few ms
            // for the round-trip to complete.
        }

        // The JedisPubSub callbacks deliver (channel, pattern, message).
        // We try to JSON-decode the payload so users see structured data
        // when the publisher used the hub's publish() helper.
        private void dispatch(String channel, String pattern, String message) {
            Object payload;
            try {
                payload = JsonUtil.parseAny(message);
            } catch (RuntimeException e) {
                payload = message;
            }
            ReceivedMessage rm = new ReceivedMessage(
                    channel == null ? "" : channel,
                    pattern,
                    payload,
                    System.currentTimeMillis()
            );
            synchronized (bufferLock) {
                buffer.addFirst(rm);
                while (buffer.size() > bufferSize) {
                    buffer.removeLast();
                }
                received++;
            }
        }

        public String name() { return name; }

        public List<String> targets() { return targets; }

        public boolean isPattern() { return isPattern; }

        public long receivedTotal() {
            synchronized (bufferLock) {
                return received;
            }
        }

        public List<ReceivedMessage> messages(int limit) {
            synchronized (bufferLock) {
                List<ReceivedMessage> out = new ArrayList<>(buffer);
                if (limit >= 0 && out.size() > limit) {
                    out = out.subList(0, limit);
                }
                return new ArrayList<>(out);
            }
        }

        public void resetReceived() {
            synchronized (bufferLock) {
                buffer.clear();
                received = 0;
            }
        }

        /**
         * Stop the subscriber thread cleanly.
         *
         * <p>Jedis ships {@code UNSUBSCRIBE}/{@code PUNSUBSCRIBE} back over
         * the same socket; after the listener has unsubscribed from every
         * target the blocking {@code subscribe}/{@code psubscribe} call
         * returns and the thread exits. As a belt-and-braces guard we also
         * close the {@link Jedis} connection in the {@code finally} block
         * of the worker thread, which unblocks any read still in flight.</p>
         */
        public void close() {
            if (closed) {
                return;
            }
            closed = true;
            try {
                if (listener.isSubscribed()) {
                    if (isPattern) {
                        listener.punsubscribe();
                    } else {
                        listener.unsubscribe();
                    }
                }
            } catch (Exception ignored) {
            }
            try {
                thread.join(1000);
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
            if (thread.isAlive()) {
                // Last resort: dropping the connection forces the
                // blocking subscribe() call to return.
                try {
                    jedis.close();
                } catch (Exception ignored) {
                }
                try {
                    thread.join(500);
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                }
            }
        }

        public boolean isAlive() {
            return !closed && thread.isAlive();
        }

        public Map<String, Object> toMap() {
            Map<String, Object> out = new LinkedHashMap<>();
            out.put("name", name);
            out.put("targets", new ArrayList<>(targets));
            out.put("is_pattern", isPattern);
            out.put("received_total", receivedTotal());
            out.put("alive", isAlive());
            return out;
        }
    }

    private final JedisPool pool;
    private final int bufferSize;

    private final Map<String, Subscription> subscriptions = new ConcurrentHashMap<>();
    private final Object subsLock = new Object();

    private final AtomicLong publishedTotal = new AtomicLong();
    private final AtomicLong deliveredTotal = new AtomicLong();
    private final Map<String, Long> channelPublished = new ConcurrentHashMap<>();

    public RedisPubSubHub(JedisPool pool) {
        this(pool, 50);
    }

    public RedisPubSubHub(JedisPool pool, int bufferSize) {
        this.pool = pool;
        this.bufferSize = bufferSize;
    }

    /**
     * Publish {@code message} to {@code channel} and return Redis'
     * delivered count.
     *
     * <p>The message is JSON-encoded so callers can pass maps, lists, or
     * scalars without serialising on every call. The returned integer is
     * what Redis itself reports: the number of clients (direct or
     * pattern) that received the message in this call.</p>
     */
    public int publish(String channel, Object message) {
        String payload = JsonUtil.toJson(message);
        long delivered;
        try (Jedis jedis = pool.getResource()) {
            delivered = jedis.publish(channel, payload);
        }
        publishedTotal.incrementAndGet();
        deliveredTotal.addAndGet(delivered);
        channelPublished.merge(channel, 1L, Long::sum);
        return (int) delivered;
    }

    /** Register a named exact-match subscription on one or more channels. */
    public Subscription subscribe(String name, List<String> channels) {
        return register(name, channels, false);
    }

    /** Register a named pattern subscription on one or more glob patterns. */
    public Subscription psubscribe(String name, List<String> patterns) {
        return register(name, patterns, true);
    }

    private Subscription register(String name, List<String> targets, boolean isPattern) {
        synchronized (subsLock) {
            if (subscriptions.containsKey(name)) {
                throw new IllegalArgumentException("subscription named '" + name + "' already exists");
            }
            Subscription sub = new Subscription(name, targets, isPattern, bufferSize);
            subscriptions.put(name, sub);
            return sub;
        }
    }

    /** Close and remove the named subscription. Returns true if it existed. */
    public boolean unsubscribe(String name) {
        Subscription sub;
        synchronized (subsLock) {
            sub = subscriptions.remove(name);
        }
        if (sub == null) {
            return false;
        }
        sub.close();
        return true;
    }

    public List<Subscription> subscriptions() {
        synchronized (subsLock) {
            return new ArrayList<>(subscriptions.values());
        }
    }

    public Subscription getSubscription(String name) {
        return subscriptions.get(name);
    }

    /** List server-side channels with at least one subscriber (PUBSUB CHANNELS). */
    public List<String> activeChannels(String pattern) {
        try (Jedis jedis = pool.getResource()) {
            List<String> channels = new ArrayList<>(jedis.pubsubChannels(pattern == null ? "*" : pattern));
            Collections.sort(channels);
            return channels;
        }
    }

    /**
     * Count subscribers per channel (PUBSUB NUMSUB).
     *
     * <p>Reports only exact-match subscriptions; pattern subscribers are
     * counted separately via {@link #patternSubscriberCount()}.</p>
     */
    public Map<String, Long> channelSubscriberCounts(List<String> channels) {
        Map<String, Long> out = new LinkedHashMap<>();
        if (channels == null || channels.isEmpty()) {
            return out;
        }
        try (Jedis jedis = pool.getResource()) {
            Map<String, Long> result = jedis.pubsubNumSub(channels.toArray(new String[0]));
            // Preserve the input ordering so the JSON output is stable.
            for (String channel : channels) {
                out.put(channel, result.getOrDefault(channel, 0L));
            }
        }
        return out;
    }

    /** Total active pattern subscriptions across all clients (PUBSUB NUMPAT). */
    public long patternSubscriberCount() {
        try (Jedis jedis = pool.getResource()) {
            return jedis.pubsubNumPat();
        }
    }

    /** Combined publish and subscribe counters plus the current registry size. */
    public Map<String, Object> stats() {
        List<Subscription> subs = subscriptions();
        long receivedTotal = 0;
        for (Subscription sub : subs) {
            receivedTotal += sub.receivedTotal();
        }
        Map<String, Object> out = new LinkedHashMap<>();
        out.put("published_total", publishedTotal.get());
        out.put("delivered_total", deliveredTotal.get());
        out.put("received_total", receivedTotal);
        out.put("active_subscriptions", (long) subs.size());
        out.put("channel_published", new LinkedHashMap<>(channelPublished));
        out.put("pattern_subscriptions", patternSubscriberCount());
        return out;
    }

    public void resetStats() {
        publishedTotal.set(0);
        deliveredTotal.set(0);
        channelPublished.clear();
        for (Subscription sub : subscriptions()) {
            sub.resetReceived();
        }
    }

    /** Close every active subscription. Safe to call more than once. */
    public void shutdown() {
        List<Subscription> subs;
        synchronized (subsLock) {
            subs = new ArrayList<>(subscriptions.values());
            subscriptions.clear();
        }
        for (Subscription sub : subs) {
            sub.close();
        }
    }
}

File: content/develop/use-cases/pub-sub/dotnet/RedisPubSubHub.cs
using System.Collections.Concurrent;
using System.Text.Json;
using System.Text.Json.Serialization;
using StackExchange.Redis;

namespace PubSubDemo;

/// <summary>
/// Message shape delivered to every <see cref="Subscription"/> callback.
/// Pattern subscriptions carry the original pattern in <see cref="Pattern"/>;
/// exact-match subscriptions leave it null.
/// </summary>
public sealed class ReceivedMessage
{
    [JsonPropertyName("channel")]
    public string Channel { get; }

    [JsonPropertyName("pattern")]
    public string? Pattern { get; }

    [JsonPropertyName("payload")]
    public object? Payload { get; }

    [JsonPropertyName("received_at_ms")]
    public long ReceivedAtMs { get; }

    public ReceivedMessage(string channel, string? pattern, object? payload, long receivedAtMs)
    {
        Channel = channel;
        Pattern = pattern;
        Payload = payload;
        ReceivedAtMs = receivedAtMs;
    }
}

/// <summary>
/// A named in-process subscriber bound to one or more channels or patterns.
///
/// StackExchange.Redis multiplexes every subscription in the process onto a
/// single underlying subscribe connection, so unlike redis-py / node-redis
/// each <see cref="Subscription"/> does NOT own its own socket. Subscriptions
/// are still independent at the application level: closing one only detaches
/// its own handler, the shared subscribe connection survives.
/// </summary>
public sealed class Subscription
{
    private readonly ISubscriber _subscriber;
    private readonly RedisChannel[] _bindings;
    private readonly Action<RedisChannel, RedisValue> _handler;
    private readonly int _bufferSize;
    private readonly LinkedList<ReceivedMessage> _buffer = new();
    private readonly object _lock = new();
    private long _received;
    private bool _closed;

    [JsonPropertyName("name")]
    public string Name { get; }

    [JsonPropertyName("targets")]
    public IReadOnlyList<string> Targets { get; }

    [JsonPropertyName("is_pattern")]
    public bool IsPattern { get; }

    [JsonPropertyName("received_total")]
    public long ReceivedTotal
    {
        get { lock (_lock) return _received; }
    }

    [JsonPropertyName("alive")]
    public bool Alive => !_closed;

    internal Subscription(
        string name,
        IEnumerable<string> targets,
        bool isPattern,
        ISubscriber subscriber,
        int bufferSize)
    {
        var targetList = targets?.ToList() ?? new List<string>();
        if (targetList.Count == 0)
        {
            throw new ArgumentException("Subscription requires at least one channel or pattern", nameof(targets));
        }
        Name = name;
        Targets = targetList;
        IsPattern = isPattern;
        _subscriber = subscriber;
        _bufferSize = bufferSize;

        _bindings = new RedisChannel[targetList.Count];
        for (var i = 0; i < targetList.Count; i++)
        {
            _bindings[i] = isPattern
                ? RedisChannel.Pattern(targetList[i])
                : RedisChannel.Literal(targetList[i]);
        }

        // The handler is stored as a field so the same delegate instance is
        // used to subscribe and unsubscribe — StackExchange.Redis matches by
        // delegate identity when detaching.
        _handler = OnMessage;
        foreach (var binding in _bindings)
        {
            _subscriber.Subscribe(binding, _handler);
        }
    }

    private void OnMessage(RedisChannel actualChannel, RedisValue value)
    {
        // For pattern subscriptions, `actualChannel` is the matched channel,
        // not the pattern. We capture the registered pattern (Targets[0] when
        // there's one binding, or `null` when multiple patterns could match
        // — we report the first that matches by glob).
        string? pattern = null;
        if (IsPattern)
        {
            var name = (string)actualChannel!;
            pattern = MatchPattern(name) ?? Targets[0];
        }

        var data = (string?)value;
        object? payload;
        try
        {
            payload = data is null ? null : JsonSerializer.Deserialize<JsonElement>(data);
        }
        catch (JsonException)
        {
            payload = data;
        }

        var message = new ReceivedMessage(
            channel: (string?)actualChannel ?? string.Empty,
            pattern: pattern,
            payload: payload,
            receivedAtMs: DateTimeOffset.UtcNow.ToUnixTimeMilliseconds()
        );

        lock (_lock)
        {
            _buffer.AddFirst(message);
            if (_buffer.Count > _bufferSize)
            {
                _buffer.RemoveLast();
            }
            _received += 1;
        }
    }

    private string? MatchPattern(string channelName)
    {
        // Cheap glob check so multi-pattern subscriptions can carry the
        // right pattern through to the buffer. Supports '*', '?' and
        // '[set]' the way Redis does (good enough for the demo UI).
        foreach (var target in Targets)
        {
            if (Glob.Match(target, channelName)) return target;
        }
        return null;
    }

    public IReadOnlyList<ReceivedMessage> Messages(int? limit = null)
    {
        lock (_lock)
        {
            if (limit is null || limit >= _buffer.Count)
            {
                return _buffer.ToArray();
            }
            return _buffer.Take(limit.Value).ToArray();
        }
    }

    public void ResetReceived()
    {
        lock (_lock)
        {
            _buffer.Clear();
            _received = 0;
        }
    }

    public bool IsAlive() => !_closed;

    public void Close()
    {
        if (_closed) return;
        _closed = true;
        foreach (var binding in _bindings)
        {
            try
            {
                _subscriber.Unsubscribe(binding, _handler);
            }
            catch
            {
                // Best effort during teardown.
            }
        }
    }
}

/// <summary>
/// Publish/subscribe helper with publisher counters and subscriber registry.
///
/// One <see cref="ConnectionMultiplexer"/> is shared across the whole hub.
/// PUBLISH and the PUBSUB introspection commands go through the regular
/// command pipeline; SUBSCRIBE / PSUBSCRIBE share a single dedicated
/// subscribe connection managed transparently by the multiplexer.
/// </summary>
public sealed class RedisPubSubHub
{
    private readonly ConnectionMultiplexer _multiplexer;
    private readonly IDatabase _db;
    private readonly ISubscriber _subscriber;
    private readonly int _bufferSize;

    private readonly ConcurrentDictionary<string, Subscription> _subscriptions = new();
    private long _publishedTotal;
    private long _deliveredTotal;
    private readonly ConcurrentDictionary<string, long> _channelPublished = new();

    public RedisPubSubHub(ConnectionMultiplexer multiplexer, int bufferSize = 50)
    {
        _multiplexer = multiplexer ?? throw new ArgumentNullException(nameof(multiplexer));
        _db = multiplexer.GetDatabase();
        _subscriber = multiplexer.GetSubscriber();
        _bufferSize = bufferSize;
    }

    /// <summary>
    /// Publish <paramref name="message"/> to <paramref name="channel"/> and
    /// return Redis' delivered count. The message is JSON-encoded so callers
    /// can pass dicts, lists, or scalars without converting on every call.
    /// </summary>
    public long Publish(string channel, object? message)
    {
        var payload = JsonSerializer.Serialize(message);
        var delivered = _subscriber.Publish(RedisChannel.Literal(channel), payload);
        Interlocked.Increment(ref _publishedTotal);
        Interlocked.Add(ref _deliveredTotal, delivered);
        _channelPublished.AddOrUpdate(channel, 1, (_, current) => current + 1);
        return delivered;
    }

    /// <summary>Register a named exact-match subscription on one or more channels.</summary>
    public Subscription Subscribe(string name, IEnumerable<string> channels) =>
        Register(name, channels, isPattern: false);

    /// <summary>Register a named pattern subscription on one or more glob patterns.</summary>
    public Subscription PSubscribe(string name, IEnumerable<string> patterns) =>
        Register(name, patterns, isPattern: true);

    private Subscription Register(string name, IEnumerable<string> targets, bool isPattern)
    {
        var sub = new Subscription(name, targets, isPattern, _subscriber, _bufferSize);
        if (!_subscriptions.TryAdd(name, sub))
        {
            sub.Close();
            throw new ArgumentException($"subscription named '{name}' already exists", nameof(name));
        }
        return sub;
    }

    /// <summary>Close and remove the named subscription. Returns true if it existed.</summary>
    public bool Unsubscribe(string name)
    {
        if (!_subscriptions.TryRemove(name, out var sub))
        {
            return false;
        }
        sub.Close();
        return true;
    }

    public IReadOnlyList<Subscription> Subscriptions() => _subscriptions.Values.ToArray();

    public Subscription? GetSubscription(string name) =>
        _subscriptions.TryGetValue(name, out var sub) ? sub : null;

    /// <summary>List server-side channels with at least one subscriber (PUBSUB CHANNELS).</summary>
    /// <remarks>
    /// StackExchange.Redis maintains an internal <c>__Booksleeve_MasterChanged</c>
    /// subscription on its shared subscribe connection so it can react to
    /// replica/master changes; we hide it here so the UI only shows user channels.
    /// </remarks>
    public string[] ActiveChannels(string pattern = "*")
    {
        var endpoints = _multiplexer.GetEndPoints();
        if (endpoints.Length == 0) return Array.Empty<string>();
        var server = _multiplexer.GetServer(endpoints[0]);
        var channels = server.SubscriptionChannels(RedisChannel.Pattern(pattern));
        return channels
            .Select(c => (string)c!)
            .Where(c => c is not null && !c.StartsWith("__Booksleeve"))
            .OrderBy(c => c, StringComparer.Ordinal)
            .ToArray();
    }

    /// <summary>
    /// Count subscribers per channel (PUBSUB NUMSUB). Reports only exact-match
    /// subscriptions — pattern subscribers are counted separately via
    /// <see cref="PatternSubscriberCount"/>.
    /// </summary>
    public Dictionary<string, long> ChannelSubscriberCounts(IEnumerable<string> channels)
    {
        var channelArray = channels?.ToArray() ?? Array.Empty<string>();
        var result = new Dictionary<string, long>();
        if (channelArray.Length == 0) return result;

        var args = new object[channelArray.Length + 1];
        args[0] = "NUMSUB";
        for (var i = 0; i < channelArray.Length; i++)
        {
            args[i + 1] = channelArray[i];
        }

        var raw = _db.Execute("PUBSUB", args);
        // PUBSUB NUMSUB returns a flat array: [ch1, count1, ch2, count2, ...]
        if (raw.Resp2Type == ResultType.Array)
        {
            var entries = (RedisResult[])raw!;
            for (var i = 0; i + 1 < entries.Length; i += 2)
            {
                var ch = (string?)entries[i] ?? "";
                var n = (long)entries[i + 1];
                result[ch] = n;
            }
        }
        // Make sure every requested channel appears, defaulting to 0.
        foreach (var ch in channelArray)
        {
            if (!result.ContainsKey(ch)) result[ch] = 0;
        }
        return result;
    }

    /// <summary>Total active pattern subscriptions across all clients (PUBSUB NUMPAT).</summary>
    public long PatternSubscriberCount()
    {
        var raw = _db.Execute("PUBSUB", "NUMPAT");
        return (long)raw;
    }

    /// <summary>Combined publish and subscribe counters plus the current registry size.</summary>
    public Dictionary<string, object> Stats()
    {
        var subs = _subscriptions.Values.ToArray();
        var channelPublished = _channelPublished
            .ToDictionary(kv => kv.Key, kv => kv.Value);
        var receivedTotal = subs.Sum(s => s.ReceivedTotal);

        return new Dictionary<string, object>
        {
            ["published_total"] = Interlocked.Read(ref _publishedTotal),
            ["delivered_total"] = Interlocked.Read(ref _deliveredTotal),
            ["received_total"] = receivedTotal,
            ["active_subscriptions"] = (long)subs.Length,
            ["channel_published"] = channelPublished,
            ["pattern_subscriptions"] = PatternSubscriberCount(),
        };
    }

    public void ResetStats()
    {
        Interlocked.Exchange(ref _publishedTotal, 0);
        Interlocked.Exchange(ref _deliveredTotal, 0);
        _channelPublished.Clear();
        foreach (var sub in _subscriptions.Values)
        {
            sub.ResetReceived();
        }
    }

    /// <summary>Close every active subscription. Safe to call more than once.</summary>
    public void Shutdown()
    {
        var subs = _subscriptions.Values.ToArray();
        _subscriptions.Clear();
        foreach (var sub in subs)
        {
            sub.Close();
        }
    }
}

// Minimal Redis-style glob matcher (* ? and [set]). Used only so a multi-
// pattern subscription can report which of its registered patterns matched
// a published channel; the actual matching is performed by Redis.
internal static class Glob
{
    public static bool Match(string pattern, string input)
    {
        return MatchAt(pattern, 0, input, 0);
    }

    private static bool MatchAt(string p, int pi, string s, int si)
    {
        while (pi < p.Length)
        {
            var c = p[pi];
            if (c == '*')
            {
                while (pi + 1 < p.Length && p[pi + 1] == '*') pi++;
                if (pi == p.Length - 1) return true;
                for (var k = si; k <= s.Length; k++)
                {
                    if (MatchAt(p, pi + 1, s, k)) return true;
                }
                return false;
            }
            if (si >= s.Length) return false;
            if (c == '?')
            {
                pi++; si++; continue;
            }
            if (c == '[')
            {
                var close = p.IndexOf(']', pi + 1);
                if (close < 0) return false;
                var set = p.Substring(pi + 1, close - pi - 1);
                if (!set.Contains(s[si])) return false;
                pi = close + 1; si++; continue;
            }
            if (c != s[si]) return false;
            pi++; si++;
        }
        return si == s.Length;
    }
}
